daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala

```diff
@@ -471,6 +472,37 @@ object Transaction {
       byKeyNodes: ImmArray[Value.NodeId],
   )
 
+  sealed abstract class DiscriminatedSubtype[X] {
+    type T <: X
+    def apply(x: X): T
+    def subst[F[_]](fx: F[X]): F[T]
+  }
+
+  object DiscriminatedSubtype {
+    def apply[X]: DiscriminatedSubtype[X] = new DiscriminatedSubtype[X] {
+      override type T = X
+      override def apply(x: X): T = x
+      override def subst[F[_]](fx: F[X]): F[T] = fx
+    }
+  }
+
+  val SubmittedTransaction = DiscriminatedSubtype[Transaction]
+  type SubmittedTransaction = SubmittedTransaction.T
+
+  val CommittedTransaction = DiscriminatedSubtype[Transaction]
+  type CommittedTransaction = CommittedTransaction.T
+
+  def commitTransaction(tx: SubmittedTransaction): CommittedTransaction =
+    CommittedTransaction(tx)
+
+  def commitTransaction(
+      tx: SubmittedTransaction,
+      f: crypto.Hash => Bytes,
+  ): Either[String, CommittedTransaction] =
+    tx.suffixCid(f).map(CommittedTransaction(_))
+
+  sealed trait CommandIdTag
```

Comment:
`CommandIdTag` doesn't seem to be used elsewhere in this PR. Is this a left-over?

Reply:
good catch.